### ISSUE TITLE
Filtering on Annotation Results is Very Slow

### ISSUE THREAD

**Issue by csaroff** [heartexlabs/label-studio#1710]
**Describe the bug**
Filtering on annotation results in labelstudio is really slow and sometimes times out.

![image](https://user-images.githubusercontent.com/7225723/140834206-eeab7491-9f35-48d7-a45c-bd8ef2244451.png)

On a dataset of ~10k tasks with only "Choice" annotations, running this filter takes ~20 seconds.

Example annotation result:

```json
[{"id": "n7JrgyJ0B8", "type": "choices", "value": {"choices": ["abc"]}, "to_name": "image", "from_name": "redacted_1"}, {"id": "mL0b2WOgeD", "type": "choices", "value": {"choices": ["def"]}, "to_name": "image", "from_name": "redacted_2"}]
```

I took a look at the underlying query in google cloud query insights

```sql
SELECT
  COUNT(*) AS "__count"
FROM
  "task_completion"
WHERE
  ("task_completion"."task_id" IN (
    SELECT
      U0."id"
    FROM
      "task" U0
    LEFT OUTER JOIN
      "task_completion" U2
    ON
      (U0."id" = U2."task_id")
    INNER JOIN
      "task_completion" U3
    ON
      (U0."id" = U3."task_id")
    INNER JOIN
      "task_completion" U4
    ON
      (U0."id" = U4."task_id")
    WHERE
      (U0."project_id" = $1
        AND UPPER(U3."result"::text) LIKE UPPER($2)
        AND UPPER(U4."result"::text) LIKE UPPER($3))
    GROUP BY
      U0."id")
    AND NOT "task_completion"."was_cancelled")
```

 After playing around with it, it looks like really slow bit is where it does a full text search on the annotation results.

```sql
        AND UPPER(U3."result"::text) LIKE UPPER('%"abc"%')
        AND UPPER(U4."result"::text) LIKE UPPER('%"def"%'))
```

For this dataset particular dataset(10k tasks) the filter takes 20 seconds and the annotation results aren't even that large.

![image](https://user-images.githubusercontent.com/7225723/140834744-a1d4490b-4b2b-4653-ad08-bb635e44c8b2.png)

But for tasks where we have brushlabel annotations, the annotation results are huge and the corresponding filters always time out.

![image](https://user-images.githubusercontent.com/7225723/140833564-ff242ecc-34ff-43a2-a957-83a913d2d7a7.png)

Is there any way to speed this query up?

The options I see to speed this up are:
1. Some clever query optimization
2. Adjust the task filtering to allow for some sort of jsonpath style querying rather than a full text search.
3. Restructure how annotation results are stored.

Thoughts?

**To Reproduce**
1. Create a dataset with ~10k tasks and corresponding "Choice" annotations.
2. Click on "Filters" and select "Where annotation results contains <foo>" 

**Expected behavior**
The filter results return within a few seconds.

**Environment (please complete the following information):**
 - Client OS: macOS Big Sur
 - Client Browser: Brave | 1.31.87 Chromium: 95.0.4638.54 (Official Build) (x86_64)
 - Label Studio Version 1.3.post1

**Comment by makseq** [comment]
@csaroff What DB do you use? 

We are working on special columns for labels and choices, hopefully this issue will be fixed by them.

**Comment by csaroff** [comment]
@makseq thats great!

We’re currently using Postgres as our database

**Comment by csaroff** [comment]
@makseq can you explain the won't fix label? 

All my filters on the annotation results where the annotation results contain segmentation annotations time out. 😞 

We have workflows that are dependent on filtering on annotation results so I'm very interested in what the path forward is here and whether or not there are any workarounds.

**Comment by makseq** [comment]
@csaroff We want to implement a normal way to filter over labels. 

![image](https://user-images.githubusercontent.com/501780/141873060-9484e594-a624-4910-b848-db81d9c68322.png)

**Comment by csaroff** [comment]
Gotcha, thanks for clarifying. Yeah, that looks much nicer!

**Comment by makseq** [comment]
@csaroff Probably in the first quarter of 2022.

**Comment by ijmiller2** [comment]
I am experience similar issues with timeouts when trying to filter on Prediction Results. In our case we have almost 100k tasks in an image classification project and filtering by predictions is very slow / basically impossible. It just keeps timing out. 

@makseq, wondering if this functionality will extend to the "Prediction Results" field and if the timeline for the release of filter by annotations is still anticipated in the near term?

**Comment by makseq** [comment]
Could you please switch to this branch and check its performance?
https://github.com/heartexlabs/label-studio/pull/2267

Yes, we are planning to add more convenient mechanics for annotation/prediction label filtering.

**Comment by makseq** [comment]
@ijmiller2 are you in cloud? why don't you contact us in the slack channel?

**Comment by ijmiller2** [comment]
@makseq, we do have the enterprise account, but we have sort a special license agreement (we're an early stage startup) that might not include the dedicated slack channel. At least we have not been added there yet, so have been relying on the community slack channel and github.

Okay, thanks for the info re: timeline projection. Appreciate the support here.

**Comment by makseq** [comment]
@ijmiller2 in this case better to write to label studio slack: 
https://label-studio.slack.com/
https://slack.labelstudio.heartex.com/?source=site - invite link

**Comment by csaroff** [comment]
@makseq Do you have any kind of an updated timeline?

I should add that storing brushlabels as images in this RLE format inside a json object that's in a list is causing performance issues that are virtually impossible to workaround.

I totally understand that migrating existing data to a new format is going to be a major challenge, but from my perspective, this is the most problematic issue with labelstudio today

**Comment by makseq** [comment]
@csaroff We did some performance optimizations [here](https://github.com/heartexlabs/label-studio/pull/3454) and [here](https://github.com/heartexlabs/label-studio/pull/3467). Now the annotation result filter works really fast in the most of our SaaS projects.

**Comment by csaroff** [comment]
@makseq thanks for working on this. Looking forward to trying this out once these changes land in [stable release](https://hub.docker.com/layers/heartexlabs/label-studio/1.7.0/images/sha256-e5c4a44a45ce4fc5b87bd278ffc0a665b13c81cc8908305d61cf8642c45bb1d4?context=explore) of the docker image.

Do you have a sense of how much performance improved for projects with a large number of brushlabels(on the order of a million images)?

**Comment by makseq** [comment]
I've checked it on projects with 100k-500k tasks, it worked almost immediately.
